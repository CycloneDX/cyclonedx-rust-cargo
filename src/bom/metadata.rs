use std::{io, marker::PhantomData};

use cargo::core::Package;
use chrono::{DateTime, Utc};
use serde::Serialize;
use xml_writer::XmlWriter;

use crate::ToXml;

#[derive(Serialize)]
pub struct Metadata<'a> {
    timestamp: DateTime<Utc>,
    #[serde(skip)]
    temp: PhantomData<&'a ()>,
}

impl<'a> Default for Metadata<'a> {
    fn default() -> Self {
        Self {
            timestamp: Utc::now(),
            temp: PhantomData::default(),
        }
    }
}

impl<'a> From<&'a Package> for Metadata<'a> {
    fn from(_pkg: &'a Package) -> Self {
        Default::default()
    }
}

impl ToXml for Metadata<'_> {
    fn to_xml<W: io::Write>(&self, xml: &mut XmlWriter<W>) -> io::Result<()> {
        xml.begin_elem("metadata")?;

        xml.elem_text(
            "timestamp",
            &self
                .timestamp
                .to_rfc3339_opts(chrono::SecondsFormat::Secs, true),
        )?;

        xml.end_elem()
    }
}
