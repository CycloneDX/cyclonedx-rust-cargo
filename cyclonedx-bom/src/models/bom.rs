/*
 * This file is part of CycloneDX Rust Cargo.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

use xml::{EmitterConfig, EventReader, EventWriter, ParserConfig};

use crate::models::component::Components;
use crate::models::composition::Compositions;
use crate::models::dependency::Dependencies;
use crate::models::external_reference::ExternalReferences;
use crate::models::metadata::Metadata;
use crate::models::property::Properties;
use crate::models::service::Services;
use crate::xml::{FromXmlDocument, ToXml};

#[derive(Debug, PartialEq)]
pub struct Bom {
    pub version: u32,
    pub serial_number: Option<UrnUuid>,
    pub metadata: Option<Metadata>,
    pub components: Option<Components>,
    pub services: Option<Services>,
    pub external_references: Option<ExternalReferences>,
    pub dependencies: Option<Dependencies>,
    pub compositions: Option<Compositions>,
    pub properties: Option<Properties>,
}

impl Bom {
    pub fn parse_from_json_v1_3<R: std::io::Read>(
        mut reader: R,
    ) -> Result<Self, crate::errors::JsonReadError> {
        let bom: crate::specs::v1_3::bom::Bom = serde_json::from_reader(&mut reader)?;
        Ok(bom.into())
    }

    pub fn parse_from_xml_v1_3<R: std::io::Read>(
        reader: R,
    ) -> Result<Self, crate::errors::XmlReadError> {
        let config = ParserConfig::default().trim_whitespace(true);
        let mut event_reader = EventReader::new_with_config(reader, config);
        let bom = crate::specs::v1_3::bom::Bom::read_xml_document(&mut event_reader)?;
        Ok(bom.into())
    }

    pub fn output_as_json_v1_3<W: std::io::Write>(
        self,
        writer: &mut W,
    ) -> Result<(), crate::errors::JsonWriteError> {
        let bom: crate::specs::v1_3::bom::Bom = self.into();
        serde_json::to_writer_pretty(writer, &bom)?;
        Ok(())
    }

    pub fn output_as_xml_v1_3<W: std::io::Write>(
        self,
        writer: &mut W,
    ) -> Result<(), crate::errors::XmlWriteError> {
        let config = EmitterConfig::default().perform_indent(true);
        let mut event_writer = EventWriter::new_with_config(writer, config);

        let bom: crate::specs::v1_3::bom::Bom = self.into();
        bom.write_xml_element(&mut event_writer)
    }
}

impl Default for Bom {
    fn default() -> Self {
        Self {
            version: 1,
            serial_number: Some(UrnUuid(format!("urn:uuid:{}", uuid::Uuid::new_v4()))),
            metadata: None,
            components: None,
            services: None,
            external_references: None,
            dependencies: None,
            compositions: None,
            properties: None,
        }
    }
}

#[derive(Debug, PartialEq)]
pub struct UrnUuid(pub(crate) String);
