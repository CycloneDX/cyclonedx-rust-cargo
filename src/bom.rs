/*
 * This file is part of CycloneDX Rust Cargo.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * Copyright (c) OWASP Foundation. All Rights Reserved.
 */
use std::io;
use std::iter::FromIterator;

use cargo::core::Package;
use serde::{Serialize, Serializer};
use uuid::Uuid;
use xml_writer::XmlWriter;

use crate::metadata::Metadata;
use crate::{Component, ToXml};

static SPEC_VERSION: &'static str = "1.3";

#[derive(Clone, Copy, Serialize)]
enum BomFormat {
    CycloneDX,
}

fn uuid_to_urn<S: Serializer>(uuid: &Uuid, serializer: S) -> Result<S::Ok, S::Error> {
    serializer.collect_str(&uuid.to_urn())
}

/// A software bill of materials for a Rust crate.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Bom {
    bom_format: BomFormat,
    spec_version: &'static str,
    #[serde(serialize_with = "uuid_to_urn")]
    serial_number: Uuid,
    version: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
    components: Vec<Component>,
}

impl<'a> Default for Bom {
    fn default() -> Bom {
        Self {
            bom_format: BomFormat::CycloneDX,
            spec_version: SPEC_VERSION,
            version: 1,
            serial_number: Uuid::new_v4(),
            metadata: None::<Metadata>,
            components: Vec::new(),
        }
    }
}

/// Create a new BOM from a sequence of cargo package references.
impl<'a> FromIterator<&'a Package> for Bom {
    fn from_iter<T: IntoIterator<Item = &'a Package>>(iter: T) -> Self {
        Self {
            components: iter.into_iter().map(Component::from).collect(),
            metadata: Some(Metadata::default()),
            ..Default::default()
        }
    }
}

impl ToXml for Bom {
    fn to_xml<W: io::Write>(&self, xml: &mut XmlWriter<W>) -> io::Result<()> {
        let namespace = format!("http://cyclonedx.org/schema/bom/{}", SPEC_VERSION);
        xml.dtd("UTF-8")?;
        xml.begin_elem("bom")?;
        xml.attr("serialNumber", &self.serial_number.to_urn().to_string())?;
        xml.attr("version", "1")?;
        xml.attr("xmlns", namespace.as_str())?;

        if let Some(metadata) = &self.metadata {
            metadata.to_xml(xml)?;
        }

        xml.begin_elem("components")?;
        for component in &self.components {
            component.to_xml(xml)?;
        }
        xml.end_elem()?;

        xml.end_elem()
    }
}
