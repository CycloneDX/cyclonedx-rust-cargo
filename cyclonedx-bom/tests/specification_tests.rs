use cyclonedx_bom::models::bom::Bom;

#[test]
fn it_should_parse_all_of_the_valid_xml_specifications() {
    insta::glob!("data/1.3/valid*.xml", |path| {
        let file = std::fs::File::open(path).expect("Failed to read file: {path:?}");
        let bom = Bom::parse_from_xml_v1_3(file)
            .expect("Failed to parse the document as an SBOM: {path:?}");

        // TODO: validate here

        let mut output = Vec::new();
        bom.output_as_xml_v1_3(&mut output)
            .expect("Failed to output the file: {path:?}");
        let bom_output = String::from_utf8_lossy(&output).to_string();

        insta::assert_snapshot!(bom_output);
    });
}

#[test]
fn it_should_parse_all_of_the_valid_json_specifications() {
    insta::glob!("data/1.3/valid*.json", |path| {
        let file = std::fs::File::open(path).expect("Failed to read file: {path:?}");
        let bom = Bom::parse_from_json_v1_3(file)
            .expect("Failed to parse the document as an SBOM: {path:?}");

        // TODO: validate here

        let mut output = Vec::new();
        bom.output_as_json_v1_3(&mut output)
            .expect("Failed to output the file: {path:?}");
        let bom_output = String::from_utf8_lossy(&output).to_string();

        insta::assert_snapshot!(bom_output);
    });
}
