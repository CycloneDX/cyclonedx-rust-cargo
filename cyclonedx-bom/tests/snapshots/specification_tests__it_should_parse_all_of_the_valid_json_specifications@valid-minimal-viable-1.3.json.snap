---
source: cyclonedx-bom/tests/specification_tests.rs
expression: bom_output
input_file: cyclonedx-bom/tests/data/1.3/valid-minimal-viable-1.3.json
---
{
  "bomFormat": "CycloneDX",
  "specVersion": "1.3",
  "version": 1,
  "serialNumber": "urn:uuid:3e671687-395b-41f5-a30f-a58921a69b79",
  "components": [
    {
      "type": "library",
      "name": "acme-library",
      "version": "1.0.0"
    }
  ]
}
